android/anc/src/main/java/org/smartregister/fhircore/anc/ui/anccare/details/AncDetailsFragment.kt

```diff
@@ -132,4 +201,99 @@ class AncDetailsFragment private constructor() : Fragment() {
   private fun populateImmunizationList(listCarePlan: List<CarePlanItem>) {
     carePlanAdapter.submitList(listCarePlan)
   }
+
+  fun buttonCQLSetOnClickListener() {
+    button_CQLEvaluate.setOnClickListener { loadCQLLibraryData() }
+  }
+
+  fun loadCQLLibraryData() {
+    ancDetailsViewModel
+      .fetchCQLLibraryData(parser, fhirResourceDataSource, LIBRARY_URL)
+      .observe(viewLifecycleOwner, this::handleCQLLibraryData)
+  }
+
+  fun loadCQLHelperData() {
+    ancDetailsViewModel
+      .fetchCQLFhirHelperData(parser, fhirResourceDataSource, HELPER_URL)
+      .observe(viewLifecycleOwner, this::handleCQLHelperData)
+  }
+
+  fun loadCQLValueSetData() {
+    ancDetailsViewModel
+      .fetchCQLValueSetData(parser, fhirResourceDataSource, VALUE_SET_URL)
+      .observe(viewLifecycleOwner, this::handleCQLValueSetData)
+  }
+
+  fun loadCQLPatientData() {
+    ancDetailsViewModel
+      .fetchCQLPatientData(parser, fhirResourceDataSource, "$PATIENT_URL$patientId/\$everything")
+      .observe(viewLifecycleOwner, this::handleCQLPatientData)
+  }
+
+  fun handleCQLLibraryData(auxLibraryData: String) {
+    libraryData = auxLibraryData
+    loadCQLHelperData()
+  }
+
+  fun handleCQLHelperData(auxHelperData: String) {
+    helperData = auxHelperData
+    loadCQLValueSetData()
+  }
+
+  fun handleCQLValueSetData(auxValueSetData: String) {
+    valueSetData = auxValueSetData
+    loadCQLPatientData()
+  }
+
+  fun handleCQLPatientData(auxPatientData: String) {
+    testData = processCQLPatientBundle(auxPatientData)
+    val parameters =
+      libraryEvaluator.runCql(
+        libraryData,
+        helperData,
+        valueSetData,
+        testData,
+        evaluatorId,
+        contextCQL,
+        contextLabel
+      )
+    val jsonObject = JSONObject(parameters)
+    textView_CQLResults.text = jsonObject.toString(4)
+    textView_CQLResults.visibility = View.VISIBLE
+  }
+
+  fun processCQLPatientBundle(auxPatientData: String): String {
+    var auxPatientDataObj = JSONObject(auxPatientData)
+    var oldJSONArrayEntry = auxPatientDataObj.getJSONArray("entry")
+    var newJSONArrayEntry = JSONArray()
+    for (i in 0 until oldJSONArrayEntry.length() - 1) {
+      var resourceType =
+        oldJSONArrayEntry.getJSONObject(i).getJSONObject("resource").getString("resourceType")
+      if (i != 0 && !resourceType.equals("Patient")) {
+        newJSONArrayEntry.put(oldJSONArrayEntry.getJSONObject(i))
+      }
+    }
+
+    auxPatientDataObj.remove("entry")
+    auxPatientDataObj.put("entry", newJSONArrayEntry)
+
+    return auxPatientDataObj.toString()
+  }
```

Comment:
I think we should move this and any other non UI related logic in the class outside this view.